contracts/strategies/traderjoe/TraderJoe.sol

```diff
@@ -31,115 +45,114 @@ contract TraderJoe is UUPSUpgradeable, StrategyOwnablePausableBaseUpgradeable {
 
     function initialize(
         StrategyArgs calldata strategyArgs,
-        ITraderJoeRouter router,
-        ITraderJoeMasterChef masterChef,
-        ITraderJoePair lpToken,
-        IERC20Upgradeable joeToken
+        TraderJoeArgs calldata traderJoeArgs
     ) external initializer {
+        __checkBinIdsAndAllocations(
+            traderJoeArgs.binIds,
+            traderJoeArgs.binAllocations
+        );
+
         __UUPSUpgradeable_init();
         __StrategyOwnablePausableBaseUpgradeable_init(strategyArgs);
 
-        TraderJoeStorage storage strategyStorage = TraderJoeStorageLib
-            .getStorage();
-
-        strategyStorage.router = router;
-        strategyStorage.masterChef = masterChef;
-        strategyStorage.lpToken = lpToken;
-        strategyStorage.joeToken = joeToken;
-
-        address token0 = lpToken.token0();
-        address token1 = lpToken.token1();
-        if (token0 != address(depositToken)) {
-            strategyStorage.pairDepositToken = IERC20Upgradeable(token0);
-        } else if (token1 != address(depositToken)) {
-            strategyStorage.pairDepositToken = IERC20Upgradeable(token1);
-        } else {
-            revert InvalidTraderJoeLpToken();
-        }
-
-        ITraderJoeMasterChef.PoolInfo memory poolInfo;
-        uint256 poolLength = masterChef.poolLength();
-        bool isPoolFound = false;
-        for (uint256 i = 0; i < poolLength; i++) {
-            poolInfo = masterChef.poolInfo(i);
-            if (address(poolInfo.lpToken) == address(lpToken)) {
-                strategyStorage.farmId = i;
-                isPoolFound = true;
-                break;
-            }
-        }
-
-        if (!isPoolFound) {
-            revert InvalidTraderJoeLpToken();
-        }
+        __initialize(
+            traderJoeArgs,
+            SwapService(
+                strategyArgs.swapServiceProvider,
+                strategyArgs.swapServiceRouter
+            )
+        );
     }
 
-    function _authorizeUpgrade(address) internal override onlyOwner {}
-
-    function _deposit(uint256 amount, NameValuePair[] calldata)
-        internal
-        virtual
-        override
+    function reinitialize(TraderJoeArgs calldata traderJoeArgs)
```

Comment:
would it be difficult to protect this method to be only calleable by the owner? Would that work with defender's proposal plugin? If it is difficult let's do it an other time. I am just slightly worried that if we were targeted, then someone else could reinitialize the contract before us with some malicious contracts.